les/api.go

```diff
@@ -0,0 +1,176 @@
+func (api *PrivateLesServerAPI) MinimumCapacity() hexutil.Uint64 {
+	return hexutil.Uint64(api.server.minCapacity)
+}
+
+// priorityClientPool stores information about prioritized clients
+type priorityClientPool struct {
+	lock                                          sync.Mutex
+	pm                                            *ProtocolManager
+	clients                                       map[enode.ID]priorityClientInfo
+	totalCap, totalpriorityCap, totalConnectedCap uint64
+	priorityCount                                 int
+}
+
+// priorityClientInfo entries exist for all prioritized clients and currently connected free clients
+type priorityClientInfo struct {
+	cap       uint64 // zero for non-priority clients
+	connected bool
+	updateCap func(uint64)
+}
+
+// SetClientCapacity sets the priority capacity assigned to a given client.
+// If the assigned capacity is bigger than zero then connection is always
+// guaranteed. The sum of capacity assigned to priority clients can not exceed
+// the total available capacity.
+//
+// Note: assigned capacity can be changed while the client is connected with
+// immediate effect.
+func (api *PrivateLesServerAPI) SetClientCapacity(id enode.ID, cap uint64) error {
+	if cap != 0 && cap < api.server.minCapacity {
+		return ErrMinCap
+	}
+
+	api.priority.lock.Lock()
+	defer api.priority.lock.Unlock()
+
+	c := api.priority.clients[id]
+	if api.priority.totalpriorityCap+cap > api.priority.totalCap+c.cap {
+		return ErrTotalCap
+	}
+	api.priority.totalpriorityCap += cap - c.cap
+	if c.updateCap != nil && cap != 0 {
+		c.updateCap(cap)
+	}
+	if c.connected {
+		if c.cap != 0 {
+			api.priority.priorityCount--
+		}
+		if cap != 0 {
+			api.priority.priorityCount++
+		}
+		api.priority.totalConnectedCap += cap - c.cap
+		api.pm.clientPool.setConnLimit(api.pm.maxFreePeers(api.priority.priorityCount, api.priority.totalConnectedCap))
+	}
+	if c.updateCap != nil && cap == 0 {
+		c.updateCap(cap)
+	}
+	if cap != 0 || c.connected {
+		c.cap = cap
+		api.priority.clients[id] = c
+	} else {
+		delete(api.priority.clients, id)
+	}
+	return nil
+}
+
+// GetClientCapacity returns the capacity assigned to a given client
+func (api *PrivateLesServerAPI) GetClientCapacity(id enode.ID) hexutil.Uint64 {
+	api.priority.lock.Lock()
+	defer api.priority.lock.Unlock()
+
+	return hexutil.Uint64(api.priority.clients[id].cap)
+}
+
+// connect should be called when a new client is connected. The callback function
+// is called when the assigned capacity is changed while the client is connected.
+// It returns the priority capacity or zero if the client is not prioritized.
+// It also returns whether the client can be accepted.
+//
+// Note: priorityClientPool also stores a record about free clients while they are
+// connected in order to be able to assign priority to them later with the callback
+// function if necessary.
+func (v *priorityClientPool) connect(id enode.ID, updateCap func(uint64)) (uint64, bool) {
```

Comment:
Right now the `priorityClientPool` relies on the `updateCap` callback to move the client to the free client pool. I think it would be better if the pools would coordinate clients among themselves. You need a reference to `freeClientPool` in `priorityClientPool` for that. Then `updateCap` would just be updating the capacity.

Reply:
Done, I managed to removed this logic entirely from the handler. Now the pools implement the peerSetNotify interface (also have their own extended interface that also contains a setLimits call). I think it is much simpler and cleaner now.